src/tools/rbd_mirror/LeaderWatcher.cc

```diff
@@ -0,0 +1,398 @@
+
+  C_SaferCond unregister_ctx;
+  unregister_watch(&unregister_ctx);
+  int r = unregister_ctx.wait();
+  if (r < 0) {
+    derr << "error unregistering leader watcher for " << m_oid << " object: "
+	 << cpp_strerror(r) << dendl;
+  }
+
+  C_SaferCond shutdown_lock_ctx;
+  {
+    Mutex::Locker locker(m_lock);
+    assert(m_leader_lock);
+
+    if (m_leader) {
+      C_SaferCond release_lock_ctx;
+      release_leader_lock(&release_lock_ctx);
+      m_lock.Unlock();
+      release_lock_ctx.wait();
+      m_lock.Lock();
+    }
+    m_leader_lock->shut_down(&shutdown_lock_ctx);
+  }
+
+  shutdown_lock_ctx.wait();
+}
+
+void LeaderWatcher::check_leader_alive(utime_t &now, int heartbeat_interval) {
+  dout(20) << dendl;
+
+  assert(m_lock.is_locked());
+
+  if (now < m_leader_last_heartbeat + 2 * heartbeat_interval) {
+    return;
+  }
+
+  dout(0) << "no hearbeat from the leader for more than "
+          << 2 * heartbeat_interval << " sec -- reacquiring the leader lock"
+          << dendl;
+
+  acquire_leader_lock(true);
+  m_leader_last_heartbeat = now;
+  m_leader_lock_owner = {};
+}
+
+void LeaderWatcher::acquire_leader_lock(bool blacklist_on_break_lock) {
+  dout(20) << dendl;
+
+  assert(m_lock.is_locked());
+
+  if (blacklist_on_break_lock && !m_leader_lock_owner.cookie.empty()) {
+    FunctionContext *ctx = new FunctionContext(
+      [this](int r) {
+        if (r < 0 && r != -ENOENT) {
+          derr << "error beaking leader lock: " << cpp_strerror(r)  << dendl;
+          return;
+        }
+
+        Mutex::Locker locker(m_lock);
+        acquire_leader_lock(false);
+      });
+
+    m_leader_lock->break_lock(m_leader_lock_owner, true, ctx);
+    return;
+  }
+
+  Context *ctx = create_context_callback<
+    LeaderWatcher, &LeaderWatcher::handle_acquire_leader_lock>(this);
+
+  m_leader_lock->try_acquire_lock(ctx);
+}
+
+void LeaderWatcher::handle_acquire_leader_lock(int r) {
+  dout(20) << dendl;
+
+  if (r < 0) {
+    derr << "error acquiring leader lock: " << cpp_strerror(r)
+         << dendl;
+
+    FunctionContext *ctx = new FunctionContext(
+      [this](int r) {
+        Mutex::Locker locker(m_lock);
+
+        if (r < 0) {
+          derr << "error retrieving leader lock owner: " << cpp_strerror(r)
+               << dendl;
+          m_leader_lock_owner = {};
+        } else {
+          m_leader_lock_owner = m_lock_owner;
+          m_leader_last_heartbeat = ceph_clock_now();
+        }
+      });
+
+    m_leader_lock->get_lock_owner(&m_lock_owner, ctx);
+    return;
+  }
+
+  MirrorStatusWatcher *status_watcher = new MirrorStatusWatcher(m_ioctx,
+                                                                m_work_queue);
+  FunctionContext *ctx = new FunctionContext(
+    [this, status_watcher](int r) {
+      Mutex::Locker locker(m_lock);
+      if (r < 0) {
+        derr << "error initializing mirror status watcher: " << cpp_strerror(r)
+             << dendl;
+        delete status_watcher;
+        return;
```

Comment:
Would this leave the leader in a zombie state where it was the leader -- but w/o the status watcher all status reports will be "down"?